zerver/lib/user_topics.py

```diff
@@ -95,39 +98,58 @@ def set_topic_mutes(
         recipient_id = stream.recipient_id
         assert recipient_id is not None
 
-        add_topic_mute(
+        set_user_topic_visibility_policy(
             user_profile=user_profile,
             stream_id=stream.id,
             recipient_id=recipient_id,
             topic_name=topic_name,
-            date_muted=date_muted,
+            visibility_policy=UserTopic.MUTED,
+            last_updated=date_muted,
         )
 
 
-def add_topic_mute(
+@transaction.atomic
```

Comment:
Why don't we need `savepoint=False` on this `transaction.atomic`? I thought this function was called from the message edit code path -- is that outer code path not using a transaction?

Reply:
I read about savepoint and durable params. -- Yes, we need savepoint=False here.

While implementing it, [test_muted_topic_remove_invalid](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L187)  is failing.
I spent 3-4 hours debugging this, no success.

This is the traceback:
```shell
======================================================================
ERROR: test_muted_topic_remove_invalid (zerver.tests.test_user_topics.MutedTopicsTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/tests/test_user_topics.py", line 192, in test_muted_topic_remove_invalid
    result = self.api_patch(user, url, data)
  File "/srv/zulip/zerver/lib/test_classes.py", line 934, in api_patch
    return self.client_patch(
  File "/srv/zulip/zerver/lib/test_helpers.py", line 404, in wrapper
    result = f(self, url, info, **kwargs)
  File "/srv/zulip/zerver/lib/test_classes.py", line 314, in client_patch
    result = django_client.patch(url, encoded, follow=follow, secure=secure, **extra)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 920, in patch
    response = super().patch(
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 495, in patch
    return self.generic("PATCH", path, data, content_type, secure=secure, **extra)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 541, in generic
    return self.request(**r)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 810, in request
    self.check_exception(response)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/test/client.py", line 663, in check_exception
    raise exc_value
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/zulip/zerver/lib/rest.py", line 39, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/srv/zulip/zerver/lib/rest.py", line 203, in rest_dispatch
    return target_function(request, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 816, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/srv/zulip/zerver/decorator.py", line 762, in _wrapped_func_arguments
    raise err
  File "/srv/zulip/zerver/decorator.py", line 759, in _wrapped_func_arguments
    return view_func(request, user_profile, *args, **kwargs)
  File "/srv/zulip/zerver/lib/request.py", line 458, in _wrapped_req_func
    return req_func(request, *args, **kwargs)
  File "/srv/zulip/zerver/views/user_topics.py", line 83, in update_muted_topic
    unmute_topic(
  File "/srv/zulip/zerver/views/user_topics.py", line 56, in unmute_topic
    stream = access_stream_for_unmute_topic_by_id(user_profile, stream_id, error)
  File "/srv/zulip/zerver/lib/streams.py", line 523, in access_stream_for_unmute_topic_by_id
    stream = Stream.objects.get(id=stream_id, realm_id=user_profile.realm_id)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 646, in get
    num = len(clone)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 376, in __len__
    self._fetch_all()
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 1867, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/query.py", line 87, in __iter__
    results = compiler.execute_sql(
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1398, in execute_sql
    cursor.execute(sql, params)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/utils.py", line 83, in _execute
    self.db.validate_no_broken_transaction()
  File "/srv/zulip-py3-venv/lib/python3.8/site-packages/django/db/backends/base/base.py", line 520, in validate_no_broken_transaction
    raise TransactionManagementError(
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
```

Reply:
**Update: The issue is resolved.**

**Still it needs discussion:**

The error was due to the order of execution of code in the test.

When this piece of code executes earlier in [test_muted_topic_remove_invalid](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L187)
```
        data = {"stream": stream.name, "topic": "BOGUS", "op": "remove"}
        result = self.api_patch(user, url, data)
        self.assert_json_error(result, "Topic is not muted")
```
It results in the above mentioned error (traceback posted in the previous comment).

**Putting it at the bottom of the test method, resolves the issue.**

**Reason:**
The mentioned piece of code  in the test calls `set_user_topic_visibility_policy_in_database`, which on failure rollbacks the test transaction (because savepoint=False).

Now, when the next piece of code in the test method executes, it leads to the error:
> django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.}}

There are two approaches to resolve this:
1. Placing it to the bottom resolves the issue, because there won't be any further query execution. (Now, rollback doesn't affect anything.)
2. We can update the code to:
```python
# Here savepoint is True
with transaction.atomic():
    data = {"stream": stream.name, "topic": "BOGUS", "op": "remove"}
    result = self.api_patch(user, url, data)
    self.assert_json_error(result, "Topic is not muted")
```

I have used the First approach because of its simplicity. Second approach adds another nested transaction within test.

Reply:
I think the second approach you describe might be better, since someone adding tests in this file in the future would not be affected by the presence of this test (and they would also know what exactly to do if they are in a similar situation).
In any case, whichever approach we go with, I think we should certainly add a comment explaining why this test needs special attention.

Reply:
It looks like you went with the second approach, with a comment after all --reading the test this lgtm.